From 4634b2c9e8e440d757af0213d98d0f6886e94f43 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Fri, 21 Jan 2022 22:05:44 +0200
Subject: [PATCH 2/2] add link to Sandboxed Google Play settings

add Sandboxed Google Play link for the work profile

The link is now dynamic and becomes hidden if GMS Core is uninstalled or
disabled.

Change-Id: I845210484611b27bf1214ded228050479f4de95d
---
 res/values/strings.xml                        |  3 ++
 res/xml/apps.xml                              | 13 +++++
 .../applications/GmsCompatAppController.java  | 48 +++++++++++++++++++
 3 files changed, 64 insertions(+)
 create mode 100644 src/com/android/settings/applications/GmsCompatAppController.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index d82da4a2ddd..f855a298369 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -12636,4 +12636,7 @@
 
     <string name="user_settings_send_censored_notifications_to_current">Send notifications to current user</string>
     <string name="user_settings_send_censored_notifications_to_current_summary">Your lock screen notifications will be forwarded to the current user if you are active in the background. Only the user\'s name, the app\'s name, and the time received will be shown.</string>
+
+    <string name="sandboxed_google_play">Sandboxed Google Play</string>
+    <string name="sandboxed_google_play_work_profile">Sandboxed Google Play (work profile)</string>
 </resources>
diff --git a/res/xml/apps.xml b/res/xml/apps.xml
index c4313f645e4..440ad50b6ec 100644
--- a/res/xml/apps.xml
+++ b/res/xml/apps.xml
@@ -126,4 +126,17 @@
         </Preference>
     </PreferenceCategory>
 
+    <Preference
+        android:key="sandboxed_google_play"
+        android:title="@string/sandboxed_google_play"
+        android:order="-980"
+        settings:controller="com.android.settings.applications.GmsCompatAppController"/>
+
+    <Preference
+        android:key="sandboxed_google_play_work_profile"
+        android:title="@string/sandboxed_google_play_work_profile"
+        android:order="-979"
+        settings:forWork="true"
+        settings:controller="com.android.settings.applications.GmsCompatAppController"/>
+
 </PreferenceScreen>
diff --git a/src/com/android/settings/applications/GmsCompatAppController.java b/src/com/android/settings/applications/GmsCompatAppController.java
new file mode 100644
index 00000000000..32fe7d05e60
--- /dev/null
+++ b/src/com/android/settings/applications/GmsCompatAppController.java
@@ -0,0 +1,48 @@
+package com.android.settings.applications;
+
+import android.app.compat.gms.GmsCompat;
+import android.content.Context;
+import android.content.Intent;
+import android.os.UserHandle;
+import android.text.TextUtils;
+
+import androidx.preference.Preference;
+
+import com.android.internal.gmscompat.GmsCompatApp;
+import com.android.internal.gmscompat.GmsInfo;
+import com.android.settings.core.BasePreferenceController;
+
+public class GmsCompatAppController extends BasePreferenceController {
+
+    public GmsCompatAppController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        UserHandle workProfile = getWorkProfileUser();
+        int userId = workProfile != null ?
+                workProfile.getIdentifier() :
+                UserHandle.myUserId();
+
+        return GmsCompat.isEnabledFor(GmsInfo.PACKAGE_GMS_CORE, userId) ?
+                AVAILABLE : DISABLED_FOR_USER;
+    }
+
+    @Override
+    public boolean handlePreferenceTreeClick(Preference preference) {
+        if (!TextUtils.equals(preference.getKey(), getPreferenceKey())) {
+            return false;
+        }
+        Intent intent = new Intent(GmsCompatApp.PKG_NAME + ".SETTINGS_LINK");
+        intent.setPackage(GmsCompatApp.PKG_NAME);
+
+        UserHandle workProfile = getWorkProfileUser();
+        if (workProfile != null) {
+            mContext.startActivityAsUser(intent, workProfile);
+        } else {
+            mContext.startActivity(intent);
+        }
+        return true;
+    }
+}
-- 
2.34.1

